File: bench/langs/ledger.dart
/// Keeps the running balance for one account.
class Ledger {
  static const int limit = 5000;

  final String owner;

  Ledger(this.owner);

  /// Bills the account and returns what is left.
  int charge(int amount) {
    if (amount > limit) {
      return refuse(amount);
    }
    return commitEntry(amount);
  }

  int refuse(int amount) {
    warnOwner(owner);
    return 0;
  }

  int commitEntry(int amount) {
    writeEntry(owner, amount);
    return amount;
  }
}

File: bench/langs/Ledger.hs
-- | Keeps the running balance for one account.
module Bench.Ledger (charge) where

limit :: Int
limit = 5000

-- | Bills the account and returns what is left.
charge :: String -> Int -> Int
charge owner amount =
  if amount > limit
    then refuse owner amount
    else commitEntry owner amount

refuse :: String -> Int -> Int
refuse owner _ = warnOwner owner

commitEntry :: String -> Int -> Int
commitEntry owner amount = writeEntry owner amount
